### PR TITLE
SLIP: Switch to poll based design

### PR DESCRIPTION
**Pull request by michi-jung**
## Summary

This is a refactored version of the SLIP network driver.  Updates include:

1. The original design started two kernel threads per SLIP device. The refactored version uses file_poll to essentially be driven by the UART RX and TX interrupts and pushes work to the low priority work queue.

2. The SLIP byte un-/stuffing is more efficient now, using memcpy instead of handling each byte individually.

3. The switch of the old SLIP driver to IOBs caused buffer overwrites if packets were sent that would not fit into a single IOB.  This is fixed now.

## Impact

Changes are contained to the SLIP network driver.  This fix is reducing latency, improving throughput and is more resource efficient.

## Testing

Working fine for me with LwM2M, Telnet, and iperf

### CODE REVIEW

File: drivers/net/slip.c [hunk 1]
```diff
@@ -1,4 +1,4 @@
-/****************************************************************************
+/******************************************************************************
```
Comment (xiaoxiang781216) [inline drivers/net/slip.c:1]: should we keep the max line size(79)?

Comment (michi-jung) [reply drivers/net/slip.c:1]: Ok, will do.

For reference: [NuttX C Coding Standard - Lines](https://nuttx.apache.org/docs/latest/contributing/coding_style.html#lines)

**Review by jerpelea — APPROVED** [review]

**Comment by acassis** [comment]
Very nice @michi-jung ! I want to create a video tutorial explaining how to use SLIP. My last video was about using ZMODEM, so SLIP demo should be nice! Was it easy to use/test?

**Comment by acassis** [comment]
@michi-jung did you test with Linux? http://patrickst.blogspot.com/2011/11/tcpip-over-slip-on-gnulinux-ubuntu.html

**Comment by xiaoxiang781216** [comment]
> 3. The switch of the old SLIP driver to IOBs caused buffer overwrites if packets were sent that would not fit into a single IOB.  This is fixed now.
> 

@michi-jung but the change need resolve two big buffer and copy them to/from iob. Why not simply extend iob to SLIP MTU or support the chain IOB instead?

**Comment by michi-jung** [comment]
> @michi-jung did you test with Linux? http://patrickst.blogspot.com/2011/11/tcpip-over-slip-on-gnulinux-ubuntu.html

Hi @acassis, yes, only on Debian 11 (bullseye).  To setup a point-to-point connection the following was sufficient for me (as root or with an account with sufficient capabilities):

```
slattach -L -s 115200 -p slip /dev/ttyUSB0 &
ifconfig sl0 10.0.0.1 pointtopoint 10.0.0.2 up
route add 10.0.0.2 dev sl0
```

To enable access to the network access from you devices to the wider network, you have to enable ip forwarding and network address translation on the host:

```
echo 1 > /proc/sys/net/ipv4/ip_forward
iptables -t nat -A POSTROUTING -o enp0s31f6 -j MASQUERADE
iptables -A FORWARD -i enp0s31f6 -o sl0 -m state --state RELATED,ESTABLISHED -j ACCEPT
iptables -A FORWARD -i enp0s31f6 -o sl0 -m state --state RELATED,ESTABLISHED -j ACCEPT
```

Since the MTU for SLIP is very small (296 bytes), IP packets from the internet will often be fragmented by the host before being forwarded to the device.  Thus, you should have IP defragmentation enabled in NuttX (`CONFIG_NET_IPFRAG=y`).

**Comment by michi-jung** [comment]
> @michi-jung but the change need resolve two big buffer and copy them to/from iob. Why not simply extend iob to SLIP MTU or support the chain IOB instead?

I tried to avoid the complexity of doing the byte de-/stuffing in IOBs.  And the two buffers are still smaller than the two stacks required by the tx and rx kernel threads in the old design.

However, it is of course possible to do this in IOBs.  I will open an issue to investigate this in the future.  I am currently short on time, but will do this over the next couple of weeks.

**Comment by xiaoxiang781216** [comment]
Thanks, @michi-jung

**Pull request merged**